components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/ClaimUtil.java

```diff
@@ -305,6 +316,56 @@ private static String getSpMappedRoleClaim(ServiceProvider serviceProvider,
         return OIDCClaimUtil.getServiceProviderMappedUserRoles(serviceProvider, rolesList, claimSeparator);
     }
 
+    private static List<String> filterSharedRoles(ServiceProvider serviceProvider,
+                                                  List<String> userRoles) throws FrameworkException {
+
+        List<String> rolesAssociatedWithApp = new ArrayList<>();
+        if (CollectionUtils.isNotEmpty(userRoles)) {
+            rolesAssociatedWithApp = new ArrayList<>(userRoles);
+            /*
+             Check whether the application is a fragment application and if so, add only the shared roles
+             to the user's role list.
+            */
+            if (serviceProvider.getSpProperties() != null && Arrays.stream(serviceProvider.getSpProperties()).anyMatch(
+                    spProperty -> ApplicationConstants.IS_FRAGMENT_APP.equals(spProperty.getName())
+                            && StringUtils.equals(Boolean.TRUE.toString(), spProperty.getValue()))) {
+                if (RoleConstants.ORGANIZATION.equalsIgnoreCase(serviceProvider.getAssociatedRolesConfig().
+                        getAllowedAudience())) {
+                    for (String roleName : rolesAssociatedWithApp) {
+                        try {
+                            String roleId = OAuth2ServiceComponentHolder.getInstance().getRoleManagementServiceV2().
+                                    getRoleIdByName(removeInternalDomain(roleName), RoleConstants.ORGANIZATION,
+                                            serviceProvider.getTenantDomain(), serviceProvider.getTenantDomain());
+                            if (roleId != null) {
+                                Role role = OAuth2ServiceComponentHolder.getInstance().getRoleManagementServiceV2().
```

Comment:
We may not need to get the role object in order to decide whether this is shared role or not. 
We can try to get the main role, if it exist it is a shared role. 
What need to check is whether we have db calls for both approaches or any approach is backed by cache.